crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+        })
+    }
+
+    fn instance_type(
+        defs: &[crate::InstanceType],
+        features: &WasmFeatures,
+        parents: &[ComponentState],
+        offset: usize,
+    ) -> Result<InstanceType> {
+        let mut types = ComponentTypeSpace::default();
+        let mut exports = HashMap::new();
+
+        for def in defs {
+            match def {
+                crate::InstanceType::Type(ty) => {
+                    types.type_def(ty.clone(), features, parents, offset)?;
+                }
+                crate::InstanceType::Export { name, ty } => {
+                    if exports
+                        .insert(
+                            name.to_string(),
+                            ComponentEntityType::from_type_def(
+                                types.get(*ty, offset)?.clone(),
+                                "export",
+                                offset,
+                            )?,
+                        )
+                        .is_some()
+                    {
+                        return Err(BinaryReaderError::new(
+                            format!("duplicate export name `{}` already defined", name),
+                            offset,
+                        ));
+                    }
+                }
+                crate::InstanceType::OuterType { count, index } => {
+                    let ty = if *count == 0 {
+                        types.get(*index, offset)?.clone()
+                    } else {
+                        let count = *count as usize;
+                        if count > parents.len() {
+                            return Err(BinaryReaderError::new(
+                                format!("invalid outer alias count of {}", count),
+                                offset,
+                            ));
+                        }
+
+                        parents[parents.len() - count]
+                            .types
+                            .get(*index, offset)?
+                            .clone()
+                    };
+
+                    types.0.push(ty);
+                }
+            };
+        }
+
+        Ok(InstanceType {
+            exports: Rc::new(exports),
+        })
+    }
+
+    fn function_type(
+        &self,
+        ty: crate::ComponentFuncType,
+        offset: usize,
+    ) -> Result<ComponentFuncType> {
+        let mut core_params = Vec::new();
+        let mut core_returns = Vec::new();
+
+        let params = ty
+            .params
+            .iter()
+            .map(|(name, ty)| {
+                Self::check_name(name, "function parameter", offset)?;
+                let ty = InterfaceType::new(*ty, self, offset)?;
+                ty.push_wasm_types(&mut core_params)?;
+                Ok((name.to_string(), ty))
+            })
+            .collect::<Result<_>>()?;
+
+        let result = InterfaceType::new(ty.result, self, offset)?;
+        result.push_wasm_types(&mut core_returns)?;
```

Comment:
This is a sort of "rubber needs to meet the road situation" where while it'd be great if this were the case for return types, and while I realized this is how the spec wants to work, no LLVM-based language can really produce a module that returns multiple values today. That means that the current canonical ABI implemented by wit-bindgen and such change this to a different return value depending on the context. When used for functions imported to the wasm module it's an extra parameter of where to write things and for functions exported from the wasm module it's a return pointer for where these return types are stored.

I don't know the best way to model that here but I wanted to point this out as this isn't compatible with wit-bindgen at the moment.

Reply:
You're right, this is wrong and this should be doing the same conversion from multiple returns to a single return [performed by wit-bindgen](https://github.com/bytecodealliance/wit-bindgen/blob/10e46993d0ff3ee688c5379f4223c8cf03a6f82a/crates/parser/src/abi.rs#L1074-L1089).

I'll fix this before merging this PR.

Reply:
After looking more closely at this, this isn't something easily fixed in the validator as there is no immediate context at the point where functions are lifted and lowered regarding how they would wind up in an inner module's function space.

I'm wondering if perhaps, in lieu of proper support for multivalue in needed toolsets, we can have the "componentizing tool" (also in lieu of support in lld) emit some simple adapters for the core modules to unify the import/export canonical signatures based on multivalue?

Reply:
I don't personally mind where this is handled, it'll just need to be handled somewhere is mostly what I'm pointing out. I think you're right that a native component output for LLD is probably quite far away, so in the meantime given that we'll need our own wrapper in one form or another it seems natural to do this there as well.

That being said editing the wasm binary is no small undertaking and will be quite ambitious, especially if we want to do it "correctly" and not modify the code section to preserve debuginfo. We'll need to optionally add types to the type section, edit the export section, edit the name section, etc. None of this is impossible but it's a pretty tricky thing that's relatively invasive to the module-at-hand.

Although with _that_ being said not having to deal with this in the validator/spec would be quite nice, perhaps the best state we could hope for.

Reply:
Added a check for lifting a function to match the export signature for the canonical ABI and lowering a function creates an import signature for the canonical ABI.